File: tools/rcon_status_server.py
import json
import os
import socket
import struct
import time
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from urllib.parse import urlparse


RCON_HOST = os.environ.get("RCON_HOST", "202.189.10.12")
RCON_PORT = int(os.environ.get("RCON_PORT", "42008"))
RCON_PASSWORD = os.environ.get("RCON_PASSWORD", "")
HTTP_HOST = os.environ.get("STATUS_HTTP_HOST", "127.0.0.1")
HTTP_PORT = int(os.environ.get("STATUS_HTTP_PORT", "8787"))


def make_packet(packet_id, packet_type, body):
    payload = body.encode("utf-8")
    size = 4 + 4 + len(payload) + 2
    return struct.pack("<iii", size, packet_id, packet_type) + payload + b"\x00\x00"


def read_packet(sock):
    size_data = sock.recv(4)
    if len(size_data) < 4:
        raise ConnectionError("RCON connection closed")

    size = struct.unpack("<i", size_data)[0]
    data = b""
    while len(data) < size:
        chunk = sock.recv(size - len(data))
        if not chunk:
            raise ConnectionError("RCON response ended early")
        data += chunk

    packet_id, packet_type = struct.unpack("<ii", data[:8])
    body = data[8:-2].decode("utf-8", errors="replace")
    return packet_id, packet_type, body


def read_matching(sock, expected_id, timeout=8):
    deadline = time.time() + timeout
    bodies = []

    while time.time() < deadline:
        try:
            packet_id, _, body = read_packet(sock)
        except socket.timeout:
            break

        if packet_id == -1:
            raise PermissionError("RCON authentication failed")

        if packet_id == expected_id and body:
            bodies.append(body)
            return "\n".join(bodies)

    return "\n".join(bodies)


def send_command(command):
    if not RCON_PASSWORD:
        raise RuntimeError("RCON_PASSWORD is not configured")

    with socket.create_connection((RCON_HOST, RCON_PORT), timeout=8) as sock:
        sock.settimeout(2)
        sock.sendall(make_packet(1, 3, RCON_PASSWORD))
        time.sleep(0.4)
        sock.sendall(make_packet(2, 2, command))
        body = read_matching(sock, 2, timeout=8)
        if not body:
            raise TimeoutError("RCON command returned no data")
        return body


def send_command_with_retry(command, attempts=5):
    last_error = None
    for attempt in range(attempts):
        try:
            return send_command(command)
        except Exception as exc:
            last_error = exc
            if attempt < attempts - 1:
                time.sleep(1)
    raise last_error


def parse_server_info(raw):
    try:
        info = json.loads(raw)
    except json.JSONDecodeError:
        info = {}

    player_count = int(info.get("PlayerCount_I") or 0)
    max_players = int(info.get("MaxPlayers") or 0)
    public_queue = int(info.get("PublicQueue_I") or 0)
    reserved_queue = int(info.get("ReservedQueue_I") or 0)

    return {
        "ok": True,
        "source": "RCON",
        "server": {
            "serverName": info.get("ServerName_s", ""),
            "map": info.get("MapName_s", ""),
            "gameMode": info.get("GameMode_s", ""),
            "factionA": info.get("TeamOne_s", ""),
            "factionB": info.get("TeamTwo_s", ""),
            "playerCount": player_count,
            "maxPlayers": max_players,
            "queueCount": public_queue + reserved_queue,
            "publicQueue": public_queue,
            "reservedQueue": reserved_queue,
            "nextLayer": info.get("NextLayer_s", ""),
            "playTimeSeconds": int(info.get("PLAYTIME_I") or 0),
            "source": "RCON",
        },
    }


class StatusHandler(BaseHTTPRequestHandler):
    def write_json(self, body, status=200):
        encoded = json.dumps(body, ensure_ascii=False).encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "GET, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")
        self.send_header("Cache-Control", "no-store")
        self.end_headers()
        self.wfile.write(encoded)

    def do_OPTIONS(self):
        self.write_json({"ok": True})

    def do_GET(self):
        path = urlparse(self.path).path
        if path not in ("/", "/status"):
            self.write_json({"ok": False, "error": "not found"}, 404)
            return

        try:
            raw = send_command_with_retry("ShowServerInfo")
            self.write_json(parse_server_info(raw))
        except Exception as exc:
            self.write_json({"ok": False, "error": str(exc), "source": "RCON"}, 502)

    def log_message(self, format, *args):
        return


if __name__ == "__main__":
    server = ThreadingHTTPServer((HTTP_HOST, HTTP_PORT), StatusHandler)
    print(f"RCON status server listening on http://{HTTP_HOST}:{HTTP_PORT}/status")
    server.serve_forever()

File: tools/rcon_status_server.js
const fs = require("fs");
const http = require("http");
const net = require("net");
const path = require("path");

function loadConfig() {
  const configPath = path.join(__dirname, "rcon_status_config.json");
  let fileConfig = {};

  if (fs.existsSync(configPath)) {
    fileConfig = JSON.parse(fs.readFileSync(configPath, "utf8"));
  }

  return {
    rconHost: process.env.RCON_HOST || fileConfig.rconHost || "202.189.10.12",
    rconPort: Number(process.env.RCON_PORT || fileConfig.rconPort || 42008),
    rconPassword: process.env.RCON_PASSWORD || fileConfig.rconPassword || "",
    httpHost: process.env.STATUS_HTTP_HOST || fileConfig.httpHost || "127.0.0.1",
    httpPort: Number(process.env.STATUS_HTTP_PORT || fileConfig.httpPort || 8787),
  };
}

const config = loadConfig();

function makePacket(id, type, body) {
  const bodyBuffer = Buffer.from(body, "utf8");
  const packet = Buffer.alloc(4 + 4 + 4 + bodyBuffer.length + 2);

  packet.writeInt32LE(4 + 4 + bodyBuffer.length + 2, 0);
  packet.writeInt32LE(id, 4);
  packet.writeInt32LE(type, 8);
  bodyBuffer.copy(packet, 12);
  packet.writeInt16LE(0, 12 + bodyBuffer.length);

  return packet;
}

function readPacket(socket) {
  return new Promise((resolve, reject) => {
    let buffer = Buffer.alloc(0);

    const cleanup = () => {
      socket.off("data", onData);
      socket.off("error", onError);
      socket.off("end", onEnd);
    };

    const onError = (error) => {
      cleanup();
      reject(error);
    };

    const onEnd = () => {
      cleanup();
      reject(new Error("RCON connection closed"));
    };

    const onData = (chunk) => {
      buffer = Buffer.concat([buffer, chunk]);
      if (buffer.length < 4) {
        return;
      }

      const size = buffer.readInt32LE(0);
      if (buffer.length < size + 4) {
        return;
      }

      cleanup();
      const packet = buffer.subarray(4, size + 4);
      const id = packet.readInt32LE(0);
      const body = packet.subarray(8, packet.length - 2).toString("utf8");
      resolve({ id, body });
    };

    socket.on("data", onData);
    socket.on("error", onError);
    socket.on("end", onEnd);
  });
}

function connectRcon() {
  return new Promise((resolve, reject) => {
    const socket = net.createConnection(
      { host: config.rconHost, port: config.rconPort, timeout: 8000 },
      () => resolve(socket)
    );

    socket.once("timeout", () => {
      socket.destroy();
      reject(new Error("RCON connection timed out"));
    });
    socket.once("error", reject);
  });
}

async function sendCommand(command) {
  if (!config.rconPassword) {
    throw new Error("RCON password is not configured");
  }

  const socket = await connectRcon();

  try {
    socket.write(makePacket(1, 3, config.rconPassword));
    await new Promise((resolve) => setTimeout(resolve, 400));
    socket.write(makePacket(2, 2, command));

    const deadline = Date.now() + 8000;
    while (Date.now() < deadline) {
      const packet = await readPacket(socket);
      if (packet.id === -1) {
        throw new Error("RCON authentication failed");
      }
      if (packet.id === 2 && packet.body) {
        return packet.body;
      }
    }

    throw new Error("RCON command returned no data");
  } finally {
    socket.destroy();
  }
}

async function sendCommandWithRetry(command, attempts = 5) {
  let lastError;

  for (let index = 0; index < attempts; index += 1) {
    try {
      return await sendCommand(command);
    } catch (error) {
      lastError = error;
      if (index < attempts - 1) {
        await new Promise((resolve) => setTimeout(resolve, 1000));
      }
    }
  }

  throw lastError;
}

function toInt(value) {
  const number = Number(value || 0);
  return Number.isFinite(number) ? number : 0;
}

function parseServerInfo(raw) {
  const info = JSON.parse(raw);
  const publicQueue = toInt(info.PublicQueue_I);
  const reservedQueue = toInt(info.ReservedQueue_I);

  return {
    ok: true,
    source: "RCON",
    server: {
      serverName: info.ServerName_s || "",
      map: info.MapName_s || "",
      gameMode: info.GameMode_s || "",
      factionA: info.TeamOne_s || "",
      factionB: info.TeamTwo_s || "",
      playerCount: toInt(info.PlayerCount_I),
      maxPlayers: toInt(info.MaxPlayers),
      queueCount: publicQueue + reservedQueue,
      publicQueue,
      reservedQueue,
      nextLayer: info.NextLayer_s || "",
      playTimeSeconds: toInt(info.PLAYTIME_I),
      source: "RCON",
    },
  };
}

function writeJson(response, status, body) {
  const payload = JSON.stringify(body);

  response.writeHead(status, {
    "Content-Type": "application/json; charset=utf-8",
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Methods": "GET, OPTIONS",
    "Access-Control-Allow-Headers": "Content-Type",
    "Cache-Control": "no-store",
  });
  response.end(payload);
}

const server = http.createServer(async (request, response) => {
  if (request.method === "OPTIONS") {
    writeJson(response, 200, { ok: true });
    return;
  }

  const url = new URL(request.url, `http://${request.headers.host}`);
  if (request.method !== "GET" || !["/", "/status"].includes(url.pathname)) {
    writeJson(response, 404, { ok: false, error: "not found" });
    return;
  }

  try {
    const raw = await sendCommandWithRetry("ShowServerInfo");
    writeJson(response, 200, parseServerInfo(raw));
  } catch (error) {
    writeJson(response, 502, {
      ok: false,
      source: "RCON",
      error: error instanceof Error ? error.message : String(error),
    });
  }
});

server.listen(config.httpPort, config.httpHost, () => {
  console.log(`RCON status server listening on http://${config.httpHost}:${config.httpPort}/status`);
});
